src/Libraries/Microsoft.Extensions.AI.Abstractions/AdditionalPropertiesDictionary.cs

```diff
@@ -103,6 +108,52 @@ public object? this[string key]
     /// <inheritdoc />
     public bool TryGetValue(string key, out object? value) => _dictionary.TryGetValue(key, out value);
 
-    /// <inheritdoc />
-    IEnumerator IEnumerable.GetEnumerator() => _dictionary.GetEnumerator();
+    /// <summary>Attempts to extract a typed value from the dictionary.</summary>
+    /// <typeparam name="T">Specifies the type of the value to be retrieved.</typeparam>
+    /// <param name="key">The key to locate.</param>
+    /// <param name="value">
+    /// The value retrieved from the dictionary, if found and successfully converted to the requested type;
+    /// otherwise, the default value of <typeparamref name="T"/>.
+    /// </param>
+    /// <returns>
+    /// <see langword="true"/> if a non-<see langword="null"/> value was found for <paramref name="key"/>
+    /// in the dictionary and converted to the requested type; otherwise, <see langword="false"/>.
+    /// </returns>
+    /// <remarks>
+    /// If a non-<see langword="null"/> is found for the key in the dictionary, but the value is not of the requested type but is
+    /// an <see cref="IConvertible"/> object, the method will attempt to convert the object to the requested type.
+    /// </remarks>
+    public bool TryGetValue<T>(string key, [NotNullWhen(true)] out T? value)
```

Comment:
Should we explicitly ban nullable type parameters?
```suggestion
    public bool TryGetValue<T>(string key, [NotNullWhen(true)] out T? value) where T : notnull
```

Reply:
What does that help? The parameter is already nullable in the case of reference types.

Reply:
It explicitly prevents passing nullable type parameters even if it doesn't change the nullability of the out type parameter.